### PR TITLE
feat(dup_enhancement#13): follower replica support query and update master partition config

### PR DESCRIPTION
**Pull request by Shuo-Jia**
# Ref-Issue
https://github.com/apache/incubator-pegasus/issues/892

# Change

follower need query master replica config for sending nfs target

### CODE REVIEW

File: src/replica/duplication/replica_follower.cpp [hunk 1]
```diff
@@ -104,27 +105,77 @@ void replica_follower::async_duplicate_checkpoint_from_master_replica()
...
-// todo(jiashuo1)
-void replica_follower::update_master_replica_config_callback(
-    error_code err, configuration_query_by_index_response &&resp)
+error_code
+replica_follower::update_master_replica_config(error_code err,
+                                               configuration_query_by_index_response &&resp)
 {
-    copy_master_replica_checkpoint(_master_replica_config.primary, _master_replica_config.pid);
+    error_code err_code = err != ERR_OK ? err : resp.err;
+    if (err_code != ERR_OK) {
```
Comment (levy5307) [inline src/replica/duplication/replica_follower.cpp:120]: dsn_unlikely

File: src/replica/duplication/replica_follower.cpp [hunk 2]
```diff
@@ -104,27 +105,77 @@ void replica_follower::async_duplicate_checkpoint_from_master_replica()
...
+                          copy_master_replica_checkpoint();
+                      }
                   });
               });
 }
 
-// todo(jiashuo1)
-void replica_follower::update_master_replica_config_callback(
-    error_code err, configuration_query_by_index_response &&resp)
+error_code
```
Comment (hycdong) [inline src/replica/duplication/replica_follower.cpp:115]: I have a suggestion that you can add the thread pool those functions running on.

File: src/replica/duplication/replica_follower.cpp [hunk 3]
```diff
@@ -104,27 +105,77 @@ void replica_follower::async_duplicate_checkpoint_from_master_replica()
...
+    _master_replica_config = resp.partitions[0];
+    return ERR_OK;
 }
 
-// todo(jiashuo1)
-void replica_follower::copy_master_replica_checkpoint(const rpc_address &node, const gpid &pid)
+void replica_follower::copy_master_replica_checkpoint()
 {
-    query_last_checkpoint_info_callback(ERR_OK, learn_response() /* todo(jiashuo1): placeholder */);
+    if (_master_replica_config.primary == rpc_address::s_invalid_address ||
```
Comment (hycdong) [inline src/replica/duplication/replica_follower.cpp:156]: I think check if primary address of master replica is invalid can move into function `update_master_replica_config`.
When master replica is under reconfiguration, it primary address is invalid. Same as the partition index check.

File: src/replica/duplication/replica_follower.cpp (hunk 2 above)
Comment (Shuo-Jia) [reply src/replica/duplication/replica_follower.cpp:119]: done

**Review by hycdong — APPROVED** [review]

**Review by levy5307 — APPROVED** [review]

**Pull request merged**